jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala

```diff
@@ -737,9 +739,96 @@ object XGBoost extends Serializable {
+      hasGroup: Boolean,
+      evalData: Map[String, Dataset[_]]): (Booster, Map[String, Array[Float]]) = {
+
+    logger.info(s"Running XGBoost ${spark.VERSION} with parameters:\n${params.mkString("\n")}")
+
+    val sc = trainingData.sparkSession.sparkContext
+    val xgbParamsFactory = new XGBoostExecutionParamsFactory(params, sc)
+    val xgbExecParams = xgbParamsFactory.buildXGBRuntimeParams
+
+    val prevBooster = xgbExecParams.checkpointParam.map { checkpointParam =>
+      val checkpointManager = new ExternalCheckpointManager(
+        checkpointParam.checkpointPath,
+        FileSystem.get(sc.hadoopConfiguration))
+      checkpointManager.cleanUpHigherVersions(xgbExecParams.numRounds)
+      checkpointManager.loadCheckpointAsScalaBooster()
+    }.orNull
+
+    val trainImpl = chooseTrainImpl(hasGroup, xgbExecParams, evalData)
+
+    try {
+      // Train for every ${savingRound} rounds and save the partially completed booster
+      val tracker = startTracker(xgbExecParams.numWorkers, xgbExecParams.trackerConf)
+      val (booster, metrics) = try {
+        val parallelismTracker = new SparkParallelismTracker(sc,
+          xgbExecParams.timeoutRequestWorkers,
+          xgbExecParams.numWorkers)
+        val rabitEnv = tracker.getWorkerEnvs
+
+        val boostersAndMetrics = trainImpl.datasetToRDD(trainingData, dsToRddParams).mapPartitions(
+          iter => {
+            val watches = trainImpl.buildWatches(iter)
+            buildDistributedBooster(watches, xgbExecParams, rabitEnv,
+              xgbExecParams.obj, xgbExecParams.eval, prevBooster)
+        }).cache()
+
+        val sparkJobThread = new Thread() {
+          override def run() {
+            // force the job
+            boostersAndMetrics.foreachPartition(() => _)
+          }
+        }
+        sparkJobThread.setUncaughtExceptionHandler(tracker)
+        sparkJobThread.start()
+        val trackerReturnVal = parallelismTracker.execute(tracker.waitFor(0L))
+        logger.info(s"Rabit returns with exit code $trackerReturnVal")
+        val (booster, metrics) = postTrackerReturnProcessing(trackerReturnVal,
+          boostersAndMetrics, sparkJobThread)
+        (booster, metrics)
+      } finally {
+        tracker.stop()
+      }
+      // we should delete the checkpoint directory after a successful training
+      xgbExecParams.checkpointParam.foreach {
+        cpParam =>
+          if (!xgbExecParams.checkpointParam.get.skipCleanCheckpoint) {
+            val checkpointManager = new ExternalCheckpointManager(
+              cpParam.checkpointPath,
+              FileSystem.get(sc.hadoopConfiguration))
+            checkpointManager.cleanPath()
+          }
+      }
+      (booster, metrics)
+    } catch {
+      case t: Throwable =>
+        // if the job was aborted due to an exception
+        logger.error("the job was aborted due to ", t)
+        sc.stop()
+        throw t
+    } finally {
+      trainImpl.cleanUpDriver()
+    }
+  }
+
```

Comment:
a lot of code here are just copy paste from the original method? 

at least we should provide some abstraction to avoid duplicate code